db/blob/blob_source.cc

```diff
@@ -0,0 +1,295 @@
+    }
+  }
+
+  return s;
+}
+
+CacheKey BlobSource::GetCacheKey(uint64_t file_number, uint64_t file_size,
+                                 uint64_t offset) const {
+  OffsetableCacheKey base_cache_key =
+      OffsetableCacheKey(db_id_, db_session_id_, file_number, file_size);
+  return base_cache_key.WithOffset(offset);
+}
+
+Status BlobSource::GetBlob(const ReadOptions& read_options,
+                           const Slice& user_key, uint64_t file_number,
+                           uint64_t offset, uint64_t file_size,
+                           uint64_t value_size,
+                           CompressionType compression_type,
+                           FilePrefetchBuffer* prefetch_buffer,
+                           PinnableSlice* value, uint64_t* bytes_read) {
+  assert(value);
+
+  const uint64_t key_size = user_key.size();
+
+  if (!IsValidBlobOffset(offset, key_size, value_size, file_size)) {
+    return Status::Corruption("Invalid blob offset");
+  }
+
+  const CacheKey cache_key = GetCacheKey(file_number, file_size, offset);
+
+  CachableEntry<Slice> blob_entry;
+
+  // TODO: We haven't support cache tiering for blob files yet, but will do it
+  // later.
+  if (blob_cache_ &&
+      lowest_used_cache_tier_ != CacheTier::kNonVolatileBlockTier) {
+    const Status s =
+        MaybeReadBlobFromCache(prefetch_buffer, read_options, cache_key, offset,
+                               true /* wait_for_cache */, &blob_entry);
+
+    if (s.ok() && blob_entry.GetValue() != nullptr) {
+      assert(blob_entry.GetValue()->size() == value_size);
+      if (bytes_read) {
+        *bytes_read = value_size;
+      }
+      value->PinSelf(*blob_entry.GetValue());
+      blob_entry.TransferTo(value);
+      return s;
+    }
+  }
+
+  assert(blob_entry.IsEmpty());
+
+  const bool no_io = read_options.read_tier == kBlockCacheTier;
+  if (no_io) {
+    return Status::Incomplete(
+        "A cache miss occurs if blob cache is enabled, and no blocking io "
+        "is allowed further.");
+  }
+
+  // Can't find the blob from the cache. Since I/O is allowed, read from the
+  // file.
+  CacheHandleGuard<BlobFileReader> blob_file_reader;
+  blob_file_cache_->GetBlobFileReader(file_number, &blob_file_reader);
+
+  assert(blob_file_reader.GetValue());
+
+  if (compression_type != blob_file_reader.GetValue()->GetCompressionType()) {
+    return Status::Corruption("Compression type mismatch when reading blob");
+  }
+
+  // Note: if verify_checksum is set, we read the entire blob record to be able
+  // to perform the verification; otherwise, we just read the blob itself. Since
+  // the offset in BlobIndex actually points to the blob value, we need to make
+  // an adjustment in the former case.
+  const uint64_t adjustment =
+      read_options.verify_checksums
+          ? BlobLogRecord::CalculateAdjustmentForRecordHeader(key_size)
+          : 0;
+  assert(offset >= adjustment);
+
+  const uint64_t record_size = value_size + adjustment;
```

Comment:
These adjustments are already made inside `BlobFileReader`; I don't think they're needed here (in fact, I think they would lead to incorrect results since `BlobFileReader` expects the offset of the blob value, not the blob header)

Reply:
Yeah, make sense. done.